keycloak-claims-mapper/src/main/java/ee/eki/tolkevarav/sso/keycloakserviceprovider/tokenclaimsmapper/ClaimsFromTolkevaravApiTokenEnricher.java

```diff
@@ -0,0 +1,189 @@
+package ee.eki.tolkevarav.sso.keycloakserviceprovider.tokenclaimsmapper;
+
+import org.apache.http.client.utils.URIBuilder;
+import org.jboss.logging.Logger;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.models.ClientSessionContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.oidc.TokenManager;
+import org.keycloak.representations.IDToken;
+import org.keycloak.utils.StringUtil;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.time.Duration;
+
+import static java.net.http.HttpResponse.BodyHandlers.ofString;
+
+class ClaimsFromTolkevaravApiTokenEnricher {
+
+    private static final String SELECTED_INSTITUTION_ID_HEADER_KEY = "X-Selected-Institution-ID";
+    private static final String TARA_SUBJECT_CLAIM_ATTRIBUTE_KEY = "tara_subject_claim";
```

Comment:
It is not certain; it needs to be configured. Just like it isn’t certain that there exists a internal service account client for generating access token – it is an assumption for the mapper to work.

> The identity provider which is used by the public client should have a standard claim-to-user-attribute mapper which saves Tara’s "sub" claim to Keycloak user attributes under the key "tara_subject_claim". From this value the mapper parses an Estonian personal identity code which can be used for identifying users in Tõlkevärav database.

I _chose_ not to use the `preferred_username` claim (like we did in previous project) because its behavior is not documented anywhere: there’s nothing to indicate that we can rely upon it to (a) be filled with the `sub` claim from Tara, (b) know in what way it’s transformed (e.g. lowercased, truncated etc). As far as I know, the fact it corresponds to the `sub` claim is not part of any API, it’s just a side-effect that may be removed in any future Keycloak version. 

On the other hand, explicitly adding the claim-to-user-attribute mapper to save Tara’s `sub` claim does not rely on some undocumented internal Keycloak logic – it is dependable and user attribute mappers are well-documented. There’s and example of the configuration of the user attribute mapper in [`realm-tolkevarav-dev.json` on line `1094`](https://github.com/keeleinstituut/tv-sso/blob/358c0960b17217ec15ac61df7b095c78ab31a9ef/keycloak-claims-mapper/src/test/resources/realm-tolkevarav-dev.json#L1094).

Reply:
In previous project we still used Keycloak's UserModel.getUserName(). My question was more of about the presence of the attribute itself and by looking at the json you mentioned, it is a requirement that such identityProviderMapper is defined. Since it is manually defined through keycloak admin console, then mapper cant always assume it strictly that it is present. Therefore having a configuration parameter for it is good. Identityprovidermapper also allows to map some field to username claim.

Also one thing to consider is that the subject may differ based on identity provider. Maybe sometime later there's a requirement to integrate google login and the mapper should be able to work then as well. Probably we need to adapt authorization-service as well, but my main point is that mapper shouldn't assume this kind of things, since it can't be sure about the configuration about rest of the keycloak

Reply:
@thenouan To clarify: Would you like met add a configuration parameter for the attribute name? Or leave it as is and move along?